### PR TITLE
Fixed OSX toolchain, add FreeBSD/musl platforms

### PR DESCRIPTION
**Pull request by staticfloat**
Lots of goodies in this one, including much better macOS C++ toolchain setup, FreeBSD and musl linux shards, as well as the products of a completely new shard buildsystem.

Many thanks to @wdv4758h and @ararslan for their help getting the FreeBSD toolchain setup!

### CODE REVIEW

File: src/UserNSRunner.jl [hunk 1]
```diff
@@ -40,6 +40,18 @@ function platform_def_mapping(platform)
...
 end
 
+"""
+    ldso_cmd(cmd::Cmd)
+
+We ship on a variety of platforms.  Some of them may not natively share the
+binary loader our sandbox is linked against.  But that's okay, because we ship
+our own.  So we'll just manually invoke it.
+"""
+function ldso_cmd(cmd::Cmd)
```
Comment (Keno) [inline src/UserNSRunner.jl]: Ideally the sandbox would be statically linked.

Comment (staticfloat) [reply src/UserNSRunner.jl]: Great idea.  Didn't realize that static linking would get rid of even the interpreter.

File: src/RootFS.jl [hunk 2]
```diff
@@ -70,34 +72,36 @@ function get_shard_hash(triplet::String = "base"; squashfs::Bool = use_squashfs)
...
+        "arm-linux-musleabihf" => "29f37b003277da768bbcd869d2a93c6588e9f3ab99e8ec96cd7c5e49cba57e3e",
+        "base" => "6248961bf057030055da502e3436025b40f27e11e410603c8fb0148ff04dcf9f",
+        "i686-linux-gnu" => "71788936b3b1e5f72a451a0974701851337a899d7ac557910240ea58917ca6c4",
+        "i686-linux-musl" => "acfcb756b7247b617b487bbe22df94f18d44e0dddc9f71c36ef311ffc9cb0367",
+        "i686-w64-mingw32" => "b04ae82fc3432a78825c57e32e86a7cb41ba9bf94fb21d75d8e0be91ef2a907c",
+        "powerpc64le-linux-gnu" => "4ba71b6e7969151189e9dfa870f48443d07b9b2cd68b0a268316067bc8a22845",
+        "x86_64-apple-darwin14" => "998408b0ba4060465b464648ca092fd5080d220b55d244c3821cbb19b5219edc",
+        "x86_64-linux-gnu" => "9dcec7f1092572eb9bddee43236fe0a9853ddba044425b01ff75ed162445f93f",
+        "x86_64-linux-musl" => "9639f19c735b426f355b073f5c2be1d6c6eb232791403bf99b3de270b8e325a3",
+        "x86_64-unknown-freebsd11.1" => "09a1ae3078c4e5f8fc43a728941674e9102f3697f78c026cf5a5725c08c742ed",
```
Comment (ararslan) [inline src/RootFS.jl]: IT'S HAPPENING!!!

File: src/Runner.jl [hunk 3]
```diff
@@ -82,12 +82,14 @@ function target_envs(target::AbstractString)
         "CC_FOR_BUILD" => "/opt/x86_64-linux-gnu/bin/gcc",
     )
 
-    # If we're on OSX, default to clang instead of gcc for CC and CXX
+    # If we're on OSX or FreeBBSD, default to clang for CC and CXX
```
Comment (ararslan) [inline src/Runner.jl]: > BBSD

Comment (staticfloat) [reply src/Runner.jl]: BBYSD

File: test/runtests.jl [hunk 4]
```diff
@@ -204,6 +204,27 @@ libfoo_script = """
     rm("$(tarball_path).sha256"; force=true)
 end
 
+if lower(get(ENV, "BINARYBUILDER_FULL_SHARD_TEST", "false") ) == "true"
```
Comment (ararslan) [inline test/runtests.jl]: Should be `lowercase` instead of `lower`

**Comment by ararslan** [comment]
I was going to `Pkg.test` this locally on FreeBSD but it looks like we'll need https://github.com/JuliaLang/METADATA.jl/pull/14310 in order for this package to even load on 0.7.

**Comment by staticfloat** [comment]
Alright, once the tests pass, I'm going to merge this, then run some tests on the packages that I build, and assuming the random smattering that I choose to run works, I'm going to cut a new BinaryBuilder release.  It's high time we had one in any case, and this is a pretty significant milestone.

**Pull request merged**